### PR TITLE
Java: Add data flow for record getters.

### PR DESCRIPTION
**Pull request by aschackmull**
This adds support for flow through Java 14 records.

### CODE REVIEW

File: java/ql/src/semmle/code/java/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -193,6 +193,18 @@ predicate readStep(Node node1, Content f, Node node2) {
     fr.getField() = f.(FieldContent).getField() and
     fr = node2.asExpr()
   )
+  or
+  exists(Record r, Method getter, Field recf, MethodAccess get |
+    getter.getDeclaringType() = r and
+    recf.getDeclaringType() = r and
+    getter.getNumberOfParameters() = 0 and
+    getter.getName() = recf.getName() and
+    not exists(getter.getBody()) and
```
Comment (aibaars) [inline java/ql/src/semmle/code/java/dataflow/internal/DataFlowPrivate.qll:202]: I suppose `abstract` records are not allowed, so checking for the absence of a body is fine, however, wouldn't it be better to use  `getter.fromSource()` instead

File: java/ql/test/library-tests/dataflow/records/A.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+public class A {
+  record Pair(Object x, Object y) { }
```
Comment (aibaars) [inline java/ql/test/library-tests/dataflow/records/A.java:2]: Could you also add some test with fields that have explicit "get" methods? I think one field with a sensible get method that just returns the field (has flow) and one silly get method that returns a constant (no flow)

**Review by aibaars — COMMENTED** [review]
Looks good to me. Could you add some more tests and perhaps use `fromSource()`?

File: java/ql/src/semmle/code/java/dataflow/internal/DataFlowPrivate.qll (hunk 1 above)
Comment (aschackmull) [reply java/ql/src/semmle/code/java/dataflow/internal/DataFlowPrivate.qll:202]: Records are implicitly final and therefore can't be abstract. It looks like checking `fromSource()` would work as well, but I didn't expect it based on the qldoc of `fromSource()` - it states:
```
   * Holds if this element pertains to a source file.
   *
   * Elements pertaining to source files may include generated elements
```
So I would actually expect `fromSource()` to hold for a generated getter when the corresponding record is `fromSource()`. I wonder why that isn't the case.

Comment (aibaars) [reply java/ql/src/semmle/code/java/dataflow/internal/DataFlowPrivate.qll:202]: Ok, sounds like it's best to keep using the empty body check.

File: java/ql/test/library-tests/dataflow/records/A.java (hunk 2 above)
Comment (aschackmull) [reply java/ql/test/library-tests/dataflow/records/A.java:2]: Done.

**Review by aibaars — APPROVED** [review]

**Pull request merged**